src/docs/references/events/index.md

```diff
@@ -0,0 +1,392 @@
+  `metadata`, etc) are treated as collection of key-value pairs and the
+  difference computation treats each top-level extended data key separately:
+  * If a key that did not exist in the extended data was added, the key will be
+    given with `null` value in the `previousValues` object.
+  * If a key had different value or was removed, the entire previous value is
+    given in the `previousValues` object, including when the value is a nested
+    data structure.
+
+The example event data above is for a `listing/updated` event produced as a
+result of an API call in the Marketplace API and the `previousValues` indicate
+that the following occurred:
+
+1. The listing `title` was updated
+2. The listing `availabilityPlan` was updated
+3. The `address` key in the listing's `publicData` was updated
+4. The `rules` key was added to the listing's `publicData`
+5. The set of listing images was updated, where 2 new images were added and one was removed
+
+Note that for all `*/deleted` events, some of the resources' attributes may
+occasionally have `null` values in the `previsouValues` object, due to stricter
+data deletion requirements.
+
+### Native and Integraiton API event data format
+
+The Integration API event data format differs from the native one as follows:
+
+* the event object itself is structured as an API resource (with `id`, `type`
+  and `attributes` keys)
+* the `resource` object is structured itself as an Integration API resource with
+  `id`, `attributes` and optionall `relationships` keys, each containing the
+  corresponding portion of resource data
+* the value of relationships attributes follows the same format as
+  [relationships normally do in the Integration
+  API](https://www.sharetribe.com/api-reference/index.html#including-related-resources)
+
+In both formats, the `previousValues` object always has the same shape as the
+`resource` object.
+
+Here are two abbreviated examples:
+
+#### Native event data format
+
+```json
+{
+  "id": "ef98e897-5b81-49a5-aca6-01d9759df075",
+  ...
+  "resource": {
+    "id": "5bbb2f6f-568f-470a-9949-a655e3f6ac46",
+    ...
+    "title": "Listing title",
+    "author": {"id": "5cf4c0eb-513f-419b-a8be-bdb6c14be10a"},
+    ...
+  },
+  "previousValues": {
+    "title": "old title"
+  }
+  ...
+}
+```
+
+#### Integration API event data format
+
+```json
+{
+  "id": "ef98e897-5b81-49a5-aca6-01d9759df075",
+  "type": "event",
+  "attributes": {
+    ...
+    "resource": {
+      "id": "5bbb2f6f-568f-470a-9949-a655e3f6ac46",
+      "type": "listing",
+      "attributes": {
+        "title": "Listing title",
+        ...
+      },
+      "relationships": {
+        "author": {"data": {"id": "5cf4c0eb-513f-419b-a8be-bdb6c14be10a", "type": "user"}},
+        ...
+      }
+    },
+    "previousValues": {
+      "attributes": {
+        "title": "old title"
+      }
+    }
+    ...
+  }
+}
+```
+
+### Event sources
+
+The following table lists all possible event sources:
+
+Source | Description
```

Comment:
Interestingly I've seen tegel as source a few times in our real event data when testing out the CLI. I'm not sure right now what is causing that and if those events should be there in the first place.

Reply:
Good catch, there are indeed almost 40k such events, all of which have been resolved as `:listing/updated`. I checked a few and they don't have `previousValues`. Looking at the code, I think that's because we detect `:listing/bookings` updates and don't ignore that attribute. That's a bug. I'll make PR to fix it and also delete those events from the db. @ovan

Reply:
Addressed in https://github.com/sharetribe/sharetribe-custom/pull/1974